sw/device/sca/lib/aes.c

```diff
@@ -0,0 +1,232 @@
+  // Each byte of the 32b input word is multiplied.
+  uint32_t t;
+  t = (uint32_t)mul2[s & 0xff] | ((uint32_t)mul2[(s >> 8) & 0xff] << 8) |
+      ((uint32_t)mul2[(s >> 16) & 0xff] << 16) |
+      ((uint32_t)mul2[(s >> 24) & 0xff] << 24);
+  return t;
+}
+
+static void aes_shift_rows(uint32_t *state) {
+  // ShiftRows on a transposed state
+  // Section 3.1 of [1]
+  state[1] = (state[1] >> 8) | (state[1] << 24);
+  state[2] = (state[2] >> 16) | (state[2] << 16);
+  state[3] = (state[3] >> 24) | (state[3] << 8);
+  return;
+}
+
+static void aes_mix_columns(uint32_t *state) {
+  // MixColumns on a transposed state
+  // Section 3.1 of [1]
+  uint32_t temp[4];
+
+  memcpy(temp, state, 16);
+
+  state[0] = temp[1] ^ temp[2] ^ temp[3];
+  state[1] = temp[0] ^ temp[2] ^ temp[3];
+  state[2] = temp[0] ^ temp[1] ^ temp[3];
+  state[3] = temp[0] ^ temp[1] ^ temp[2];
+
+  temp[0] = aes_mul2(temp[0]);
+  temp[1] = aes_mul2(temp[1]);
+  temp[2] = aes_mul2(temp[2]);
+  temp[3] = aes_mul2(temp[3]);
+
+  state[0] ^= temp[0] ^ temp[1];
+  state[1] ^= temp[1] ^ temp[2];
+  state[2] ^= temp[2] ^ temp[3];
+  state[3] ^= temp[3] ^ temp[0];
+
+  return;
+}
+
+static void aes_transpose_to_32(uint8_t *in_data, uint32_t *out_data) {
+  out_data[0] = (uint32_t)in_data[0] | ((uint32_t)in_data[4] << 8) |
+                ((uint32_t)in_data[8] << 16) | ((uint32_t)in_data[12] << 24);
+  out_data[1] = (uint32_t)in_data[1] | ((uint32_t)in_data[5] << 8) |
+                ((uint32_t)in_data[9] << 16) | ((uint32_t)in_data[13] << 24);
+  out_data[2] = (uint32_t)in_data[2] | ((uint32_t)in_data[6] << 8) |
+                ((uint32_t)in_data[10] << 16) | ((uint32_t)in_data[14] << 24);
+  out_data[3] = (uint32_t)in_data[3] | ((uint32_t)in_data[7] << 8) |
+                ((uint32_t)in_data[11] << 16) | ((uint32_t)in_data[15] << 24);
+}
+
+static void aes_transpose_from_32(uint32_t *in_data, uint8_t *out_data) {
+  out_data[0] = (uint8_t)(in_data[0] & 0xff);
+  out_data[1] = (uint8_t)(in_data[1] & 0xff);
+  out_data[2] = (uint8_t)(in_data[2] & 0xff);
+  out_data[3] = (uint8_t)(in_data[3] & 0xff);
+  out_data[4] = (uint8_t)(in_data[0] >> 8) & 0xff;
+  out_data[5] = (uint8_t)(in_data[1] >> 8) & 0xff;
+  out_data[6] = (uint8_t)(in_data[2] >> 8) & 0xff;
+  out_data[7] = (uint8_t)(in_data[3] >> 8) & 0xff;
+  out_data[8] = (uint8_t)(in_data[0] >> 16) & 0xff;
+  out_data[9] = (uint8_t)(in_data[1] >> 16) & 0xff;
+  out_data[10] = (uint8_t)(in_data[2] >> 16) & 0xff;
+  out_data[11] = (uint8_t)(in_data[3] >> 16) & 0xff;
+  out_data[12] = (uint8_t)(in_data[0] >> 24) & 0xff;
+  out_data[13] = (uint8_t)(in_data[1] >> 24) & 0xff;
+  out_data[14] = (uint8_t)(in_data[2] >> 24) & 0xff;
+  out_data[15] = (uint8_t)(in_data[3] >> 24) & 0xff;
+}
+
+static void aes_rcon_next(uint8_t *rcon) {
+  // rcon cannot be 0
+  if (*rcon) {
+    // update round constant
+    *rcon = mul2[*rcon];
+  } else {
+    // init round constant to first-round value
+    *rcon = 0x1;
+  }
+
+  return;
+}
```

Comment:
Using a pointer here doesn't bring any benefit, so I would avoid it by returning the next rcon instead.
```suggestion
static uint8_t aes_rcon_next(uint8_t rcon) {
  // rcon cannot be 0
  if (rcon != 0) {
    // update round constant
   return mul2[rcon];
  } else {
    // init round constant to first-round value
    return 0x1;
  }
}
```

Reply:
Fully agree.